src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Returns the index in the form of an integer.
+     * 
+     * @param input The index to be changed into an integer.
+     * @return Integer that represents the index.
+     * @throws DukeException If input is invalid.
+     */
+    private static int validateMarkOrDeleteIndex(String input) throws DukeException {
+        if (input.isEmpty() || isNotNumeric(input)) {
+            throw new DukeException("OOPS!! The index to edit is not valid!");
+        }
+        return Integer.parseInt(input);
+    }
+
+    /**
+     * Returns a boolean representing if the input is a number.
+     * 
+     * @param input Input to check if it is a number.
+     * @return True if input is not a number, false if otherwise.
+     */
+    private static boolean isNotNumeric(String input) {
+        return !input.matches("-?\\d+(\\.\\d+)?");
+    }
+
+    /**
+     * Checks if ToDo task is valid.
+     * 
+     * @param todoString Description of the ToDo.
+     * @throws DukeException If Description is incomplete.
+     */
+    private static void validateToDo(String todoString) throws DukeException {
+        if (todoString.isEmpty()) {
+            throw new DukeException("OOPS!! The description of a ToDo cannot be empty!");
+        }
+    }
+
+    /**
+     * Returns an array of String representing the various components of an Event.
+     * 
+     * @param input String input of the event.
+     * @return an array of String objects that represent an Event.
+     * @throws DukeException If input is invalid or missing details.
+     */
+    private static String[] validateEvent(String input) throws DukeException {
+        String description = input.replace("event ", "");
+        String[] details = description.split(" /");
+        if (details.length != 3) {
+            throw new DukeException("OOPS!! You need to include the details!");
+        }
+        details[1] = details[1].replace("from ", "");
+        details[2] = details[2].replace("to ", "");
+        if (details[1].isEmpty() || details[2].isEmpty()) {
+            throw new DukeException("OOPS! Please make sure the dates are not empty!");
+        }
+        return details;
+    }
+
+    /**
+     * Returns an array of String objects that represents a Deadline task.
+     * 
+     * @param input Input for a Deadline.
+     * @return an array of String objects that represents a Deadline.
+     * @throws DukeException If input is invalid or incomplete.
+     */
+    private static String[] validateDeadline(String input) throws DukeException {
+        if (input.equals("deadline")) {
+            throw new DukeException("OOPS!! You need to include the deadline!");
+        }
+        String description = input.replace("deadline ", "");
+        String[] details = description.split(" /by ");
+        if (details.length != 2) {
+            throw new DukeException("OOPS!! You need to include the deadline!");
+        }
+        return details;
+    }
```

Comment:
I really like how you go the extra mile to check for instances like when the user forgets to input the date!